keps/sig-node/2837-pod-level-resource-spec/README.md

```diff
@@ -0,0 +1,1458 @@
+The development environment defines a container with the web-server serving the IDE itself, and constrains it to use a certain amount of memory.
+Additional tools are provided in additional side-car deployments - for example an LSP( https://langserver.org/) service, a terminal, and more. 
+Using this new feature, the containers providing the terminal, the LSP services, and the set of tools being utilized can share the resource limit
+defined for the pod. Consider the following pod definition:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  labels:
+    run: ide
+  name: myide
+spec:
+  resources:
+    limits:
+      memory: "1024Mi"
+      cpu: "4"
+  containers:
+  - image: shell
+    name: debian:buster
+  - image: tool1
+    name: first-tool:latest
+  - image: tool2
+    name: second-tool:latest
+  - name: ide
+    image: theia:latest
+    resources:
+      requests:
+        memory: 128M
+        cpu: "0.5"
+      limits:
+        memory: 256M
+        cpu: "1"
+```
+
+Using the pod-level resource definition enables the `tool1` and `tool2` containers to be constrained by the total limit for the pod. 
+
+Without this feature, the developer would need to decide a-priori how much resources to allocate to the tools - and this is not easy to do for this workload.
+
+
+#### Story 2
+
+Consider a scenario where a pod has some containers with occasional resource spikes while other containers often have idle resources. With current container 
+level resource settings, the peak usage of each container needs to be accounted for when setting limits. This leads to wasted resources when the spikes are 
+not occurring. Pod-level resource limits will allow setting a limit for the entire pod, enabling containers to share resources more efficiently without the
+need to over-provision for each container’s peak usage. 
+
+
+### Notes/Constraints/Caveats
+
+
+1. Community is trying to move cgroupv1 to maintenance mode in future Kubernetes versions (likely 1.31). Hence this feature will be supported only on cgroupv2.
+2. This feature is an **opt-in** feature, and will have no effect on existing deployments. Only deployments that explicitly require this functionality should
+turn it on by specifying the relevant resource section at pod-level in the Pod specification.
+
+
+### Risks and Mitigations
+
+Since this is an **opt-in** feature, there should be no risk in merging the changes. Existing workloads that do not use this feature won’t be affected.
+Besides, the feature will be introduced in a phased approach, starting with an alpha release for testing and gathering feedback, followed by beta and GA releases
+as the feature matures and potential issues are addressed.
+
+
+## Design Details
+
+
+### Components/Features changes
+
+
+#### PodSpec changes
+
+New field in `PodSpec`
+
+```
+type PodSpec struct {
+...
+  // Compute resource requirements.
+  // +optional
+  Resources ResourceRequirements
+}
+```
+
+
+#### PodSpec Validation
```

Comment:
third column is concerning a bit.

I would expect that customer will be able to say:

I need that much memory for the entire Pod. And I want to limit individual containers to some high value as a precausion. But than - container requests will be equal to container limits. And sum of requests for containers will be higher than Pod request. Which will fail the pod admission as I understand

Reply:
Is it the 3rd row that you were talking about? 

I'm not sure I follow the use case. Individual container requests should represent the expected baseline of the resource consumption, and their sum should not exceed the pod requests.

If user wants to have container request == limit to have the guaranteed class behavior, they'd most likely want to have a corresponding high pod resource request and limit.

Could you elaborate more?

Reply:
Sum of requests for containers are now not allowed to be greater than pod-level requests.